lib/chain.js

```diff
@@ -0,0 +1,179 @@
+'use strict';
+
+var assert = require('assert-plus');
+var once = require('once');
+
+module.exports = Chain;
+
+/**
+ * Create a new middleware chain
+ *
+ * @public
+ * @class Chain
+ * @param {Object} [options] - options
+ * @param {Boolean} [options.onceNext=false] - Prevents calling next multiple
+ *  times
+ * @param {Boolean} [options.strictNext=false] - Throws error when next() is
+ *  called more than once, enables onceNext option
+ * @example
+ * var chain = new Chain();
+ * chain.add(function (req, res, next) { next(); })
+ * // chain.add(function (req, res, next) { next(new Error('Foo')); })
+ * // chain.add(function (req, res, next) { next(false); })
+ *
+ * http.createServer((req, res) => {
+ *    chain.run(req, res, function done(err) {
+ *       res.end(err ? err.message : 'hello world');
+ *    });
+ * })
+ */
+function Chain(options) {
+    assert.optionalObject(options, 'options');
+    options = options || {};
+    assert.optionalBool(options.onceNext, 'options.onceNext');
+    assert.optionalBool(options.strictNext, 'options.strictNext');
+
+    this.onceNext = !!options.onceNext;
+    this.strictNext = !!options.strictNext;
+
+    // strictNext next enfoces onceNext
+    if (this.strictNext) {
+        this.onceNext = true;
+    }
+
+    this._stack = [];
+    this._once = this.strictNext === false ? once : once.strict;
+}
+
+/**
+ * Public methods.
+ * @private
+ */
+
+/**
+ * Get handlers of a chain instance
+ *
+ * @memberof Chain
+ * @instance
+ * @returns {Function[]} handlers
+ */
+Chain.prototype.getHandlers = function getHandlers() {
+    return this._stack;
+};
+
+/**
+ * Utilize the given middleware `handler`
+ *
+ * @public
+ * @memberof Chain
+ * @instance
+ * @param {Function|Chain} handler - handler or Chain instance
+ * @returns {undefined} no return value
+ */
+Chain.prototype.add = function add(handler) {
+    // wrap nested chain
+    if (typeof handler.run === 'function') {
+        var chain = handler;
+        handler = function nestedChainHandler(req, res, next) {
+            chain.run(req, res, next);
+        };
+    }
+
+    // _name is assigned in the server and router
+    handler._name = handler._name || handler.name;
+
+    // add the middleware
+    this._stack.push(handler);
+};
+
+/**
+ * Returns the number of handlers
+ *
+ * @public
+ * @memberof Chain
+ * @instance
+ * @returns {Number} number of handlers in the stack
+ */
+Chain.prototype.count = function count() {
```

Comment:
Potential bikeshed - why count() vs length()/.length property? (don't feel super strongly)

Reply:
Me neither, I thought it's a good idea to differentiate it from the JavaScript Array length